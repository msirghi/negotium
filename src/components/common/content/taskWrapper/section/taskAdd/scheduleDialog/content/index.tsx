export { MenuContent } from './container/MenuContent';
