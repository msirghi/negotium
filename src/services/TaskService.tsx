import { Task } from '../common/types/tasks.types';
import { BASE_API_URL_V1 } from '../common/constants/constants';
import Requests from '../common/requests/request';
import dayjs from 'dayjs';
import { TaskOrderUpdateDto } from '../components/common/dnd/taskWrapper/types';
import { HttpMethod } from '../common/constants/enums';

const createTask = (task: Omit<Task, 'id'>) => {
  return Requests.restApiCallWithBearer(
    `${BASE_API_URL_V1}/tasks`,
    HttpMethod.POST,
    task
  );
};

const markTaskAsDone = (id: Task['id']) => {
  return Requests.restApiCallWithBearer(
    `${BASE_API_URL_V1}/tasks/${id}`,
    HttpMethod.PATCH,
    {
      completed: true,
    }
  );
};

const updateTaskName = (id: Task['id'], title: Task['title']) => {
  return Requests.restApiCallWithBearer(
    `${BASE_API_URL_V1}/tasks/${id}`,
    HttpMethod.PATCH,
    {
      title,
    }
  );
};

const updateTaskDescription = (
  id: Task['id'],
  description: Task['description']
) => {
  return Requests.restApiCallWithBearer(
    `${BASE_API_URL_V1}/tasks/${id}`,
    HttpMethod.PATCH,
    {
      description,
    }
  );
};

const updateTaskDueDate = (id: Task['id'], dueDate: Task['dueDate']) => {
  let validatedDate = null;
  if (dayjs(dueDate).isValid()) {
    validatedDate = dueDate;
  }
  return Requests.restApiCallWithBearer(
    `${BASE_API_URL_V1}/tasks/${id}`,
    HttpMethod.PATCH,
    {
      dueDate: validatedDate,
    }
  );
};

const updateOrderNumbers = (dto: TaskOrderUpdateDto) => {
  return Requests.restApiCallWithBearer(
    `${BASE_API_URL_V1}/tasks/meta/order`,
    HttpMethod.PATCH,
    dto
  );
};

const TaskService = {
  createTask,
  markTaskAsDone,
  updateTaskName,
  updateTaskDueDate,
  updateTaskDescription,
  updateOrderNumbers,
};

export default TaskService;
