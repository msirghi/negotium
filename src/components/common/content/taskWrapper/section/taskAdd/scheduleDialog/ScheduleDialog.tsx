import { FC, useState, MouseEvent } from 'react';
import { Menu } from '@mui/material';
import { MenuContent } from './content';
import DateUtils from '../../../../../../../common/utils/dateUtils';
import colors from '../../../../../../../common/styles/colors';
import { makeStyles } from '@mui/styles';
import { NullableDate } from '../../../../../../../common/types/common.types';

type Props = {
  onDateSelect: (date: NullableDate) => void;
  defaultDate?: Date;
};

const useStyles = makeStyles({
  root: {
    marginTop: '20px',
    borderRadius: 5,
    padding: '3px 5px',
    width: 'fit-content',
    cursor: 'pointer',
  },
});

export const ScheduleDialog: FC<Props> = ({ onDateSelect, defaultDate }) => {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);
  const [selectedDate, setSelectedDate] = useState<NullableDate>(
    (defaultDate || null) as unknown as NullableDate
  );
  const classes = useStyles();

  const onDatePick = (date: NullableDate) => {
    onDateSelect(date);
    setSelectedDate(date);
    handleClose();
  };

  const handleClickListItem = (event: MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <div>
      <div
        id={'schedule-button'}
        onClick={handleClickListItem}
        className={classes.root}
        style={{
          border: selectedDate
            ? `1px solid ${colors.primaries.lightBlue_1}`
            : '1px solid grey',

          color: selectedDate ? colors.primaries.lightBlue_1 : '',
        }}
      >
        <div>
          {selectedDate ? DateUtils.getDateLabel(selectedDate) : 'Schedule'}
        </div>
      </div>
      <Menu anchorEl={anchorEl} open={open} onClose={handleClose}>
        <MenuContent onDatePick={onDatePick} selectedDate={selectedDate} />
      </Menu>
    </div>
  );
};
