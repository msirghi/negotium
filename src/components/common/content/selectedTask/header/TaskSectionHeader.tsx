import { FC, useState } from 'react';
import { Box } from '@mui/system';
import { Checkbox, Divider } from '@mui/material';
import { Task } from '../../../../../common/types/tasks.types';
import { Row } from '../../../utilities/row/Row';
import { makeStyles } from '@mui/styles';
import DateUtils from '../../../../../common/utils/dateUtils';
import colors from '../../../../../common/styles/colors';
import dayjs from 'dayjs';
import CalendarTodayIcon from '@mui/icons-material/CalendarToday';
import { ScheduleDialog } from '../../taskWrapper/section/taskAdd/scheduleDialog/ScheduleDialog';
import { Nullable } from '../../../../../common/types/common.types';
import TaskService from '../../../../../services/TaskService';

type Props = {
  task: Task;
  onTaskDateUpdate: (newDate: Nullable<Date>) => void;
  markTaskAsDone: () => void;
};

const useStyles = makeStyles(() => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    width: 'fit-content',
    borderRadius: 1,
    bgcolor: 'background.paper',
    color: 'text.secondary',
    '& svg': {
      m: 1.5,
    },
    '& hr': {
      mx: 0.5,
    },
  },
  date: {
    marginLeft: 10,
    cursor: 'pointer',
  },
  dateTitle: {
    marginLeft: 10,
    marginRight: 10,
  },
  scheduleDialog: {
    padding: 5,
    borderRadius: 5,
  },
}));

export const TaskSectionHeader: FC<Props> = ({
  task,
  onTaskDateUpdate,
  markTaskAsDone,
}) => {
  const { dueDate, id } = task;
  const classes = useStyles();
  const isDateInThePast = DateUtils.isDateInThePast(dueDate);
  const [selectedDate, setSelectedDate] = useState<Nullable<Date>>(
    dueDate as unknown as Date
  );

  const onDateSelect = async (dueDate: Nullable<Date>) => {
    setSelectedDate(dueDate);
    onTaskDateUpdate(dueDate);
    if (!task.projectId) {
      await TaskService.updateTaskDueDate(
        id,
        dayjs(dueDate as unknown as Date).format()
      );
    }
  };

  return (
    <Box className={classes.root}>
      <Row alignVerticalCenter>
        <Checkbox onChange={markTaskAsDone} />
        <Divider orientation="vertical" flexItem />
        <Box
          className={classes.date}
          sx={{
            color: isDateInThePast ? colors.error.main : 'inherit',
          }}
        >
          <Row alignVerticalCenter>
            <CalendarTodayIcon color={'primary'} fontSize={'small'} />
            <Box className={classes.dateTitle}>
              <ScheduleDialog
                defaultDate={selectedDate!}
                onDateSelect={onDateSelect}
                className={classes.scheduleDialog}
              />
            </Box>
          </Row>
        </Box>
      </Row>
    </Box>
  );
};
