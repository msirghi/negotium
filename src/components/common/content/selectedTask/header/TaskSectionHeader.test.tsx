import renderer from 'react-test-renderer';
import { TaskSectionHeader } from './TaskSectionHeader';
import { TasksMock } from '../../../../../common/tests/mockData/tasks-mock';
import { mount } from 'enzyme';
import { ScheduleDialog } from '../../taskWrapper/section/taskAdd/scheduleDialog/ScheduleDialog';
import { act } from '@testing-library/react';
import TaskService from '../../../../../services/TaskService';
import DateUtils from '../../../../../common/utils/dateUtils';

describe('TaskSectionHeader', () => {
  const defaultProps = {
    onTaskDateUpdate: jest.fn(),
    task: TasksMock[0],
  };

  beforeAll(() => {
    TaskService.updateTaskDueDate = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    const tree = renderer.create(<TaskSectionHeader {...defaultProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('should match the snapshot with date not in the past', () => {
    DateUtils.isDateInThePast = jest.fn(() => false);
    const tree = renderer.create(<TaskSectionHeader {...defaultProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('should handle date update by calling prop method', () => {
    const wrapper = mount(<TaskSectionHeader {...defaultProps} />);
    const dialog = wrapper.find(ScheduleDialog);
    act(() => {
      dialog.props().onDateSelect(null);
    });
    expect(defaultProps.onTaskDateUpdate).toBeCalled();
  });
});
