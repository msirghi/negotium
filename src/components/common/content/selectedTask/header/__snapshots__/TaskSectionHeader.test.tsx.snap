// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskSectionHeader should match the snapshot 1`] = `
<div
  className="makeStyles-root-1 MuiBox-root css-0"
>
  <div
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "flex-start",
        "width": "initial",
      }
    }
  >
    <span
      className="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <input
        className="PrivateSwitchBase-input css-1m9pwf3"
        data-indeterminate={false}
        onChange={[Function]}
        type="checkbox"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CheckBoxOutlineBlankIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </span>
    <hr
      className="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
    />
    <div
      className="makeStyles-date-2 MuiBox-root css-q2y3yl"
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "justifyContent": "flex-start",
            "width": "initial",
          }
        }
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeSmall css-1bptiuh-MuiSvgIcon-root"
          data-testid="CalendarTodayIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 18H4V8h16v13z"
          />
        </svg>
        <div
          className="makeStyles-dateTitle-3 MuiBox-root css-0"
        >
          <div>
            <div
              className="makeStyles-scheduleDialog-4"
              id="schedule-button"
              onClick={[Function]}
              style={
                Object {
                  "border": "1px solid #5886ff",
                  "color": "#5886ff",
                }
              }
            >
              <div>
                12 Nov
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TaskSectionHeader should match the snapshot with date not in the past 1`] = `
<div
  className="makeStyles-root-1 MuiBox-root css-0"
>
  <div
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "flex-start",
        "width": "initial",
      }
    }
  >
    <span
      className="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <input
        className="PrivateSwitchBase-input css-1m9pwf3"
        data-indeterminate={false}
        onChange={[Function]}
        type="checkbox"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CheckBoxOutlineBlankIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </span>
    <hr
      className="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
    />
    <div
      className="makeStyles-date-2 MuiBox-root css-q2y3yl"
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "justifyContent": "flex-start",
            "width": "initial",
          }
        }
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeSmall css-1bptiuh-MuiSvgIcon-root"
          data-testid="CalendarTodayIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 18H4V8h16v13z"
          />
        </svg>
        <div
          className="makeStyles-dateTitle-3 MuiBox-root css-0"
        >
          <div>
            <div
              className="makeStyles-scheduleDialog-4"
              id="schedule-button"
              onClick={[Function]}
              style={
                Object {
                  "border": "1px solid #5886ff",
                  "color": "#5886ff",
                }
              }
            >
              <div>
                12 Nov
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
